File: cotsSource/org.apache.qpid/dotnet/Qpid.Client/Client/State/AMQStateManager.cs
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
using System;
using System.Collections;
using log4net;
using Apache.Qpid.Client.Handler;
using Apache.Qpid.Client.Protocol;
using Apache.Qpid.Client.Protocol.Listener;
using Apache.Qpid.Framing;

namespace Apache.Qpid.Client.State
{
    public class AMQStateManager : IAMQMethodListener
    {
        private static readonly ILog _logger = LogManager.GetLogger(typeof(AMQStateManager));

        const bool InfoLoggingHack = true;
            
        /// <summary>
        /// The current state
        /// </summary>
        private AMQState _currentState;

        /// <summary>
        /// Maps from an AMQState instance to a Map from Class to StateTransitionHandler.
        /// The class must be a subclass of AMQFrame.
        /// </summary>
        private readonly IDictionary _state2HandlersMap;
        private ArrayList _stateListeners;
        private object _syncLock;
        
        public AMQStateManager()
        {
            _syncLock = new object();
            _state2HandlersMap = new Hashtable();
            _stateListeners = ArrayList.Synchronized(new ArrayList(5));
            _currentState = AMQState.CONNECTION_NOT_STARTED;
            RegisterListeners();
        }

        private void RegisterListeners()
        {
            IStateAwareMethodListener connectionStart = new ConnectionStartMethodHandler();
            IStateAwareMethodListener connectionClose = new ConnectionCloseMethodHandler();
            IStateAwareMethodListener connectionCloseOk = new ConnectionCloseOkHandler();
            IStateAwareMethodListener connectionTune = new ConnectionTuneMethodHandler();
            IStateAwareMethodListener connectionSecure = new ConnectionSecureMethodHandler();
            IStateAwareMethodListener connectionOpenOk = new ConnectionOpenOkMethodHandler();
            IStateAwareMethodListener channelClose = new ChannelCloseMethodHandler();
            IStateAwareMethodListener basicDeliver = new BasicDeliverMethodHandler();
            IStateAwareMethodListener basicReturn = new BasicReturnMethodHandler();
            IStateAwareMethodListener queueDeleteOk = new QueueDeleteOkMethodHandler();
            IStateAwareMethodListener queuePurgeOk = new QueuePurgeOkMethodHandler();
            
            // We need to register a map for the null (i.e. all state) handlers otherwise you get
            // a stack overflow in the handler searching code when you present it with a frame for which
            // no handlers are registered.
            _state2HandlersMap[AMQState.ALL] = new Hashtable();

            {
                Hashtable notStarted = new Hashtable();
                notStarted[typeof(ConnectionStartBody)] = connectionStart;
                notStarted[typeof(ConnectionCloseBody)] = connectionClose;
                _state2HandlersMap[AMQState.CONNECTION_NOT_STARTED] = notStarted;
            }
            {
                Hashtable notTuned = new Hashtable();
                notTuned[typeof(ConnectionTuneBody)] = connectionTune;
                notTuned[typeof(ConnectionSecureBody)] = connectionSecure;
                notTuned[typeof(ConnectionCloseBody)] = connectionClose;
                _state2HandlersMap[AMQState.CONNECTION_NOT_TUNED] = notTuned;
            }
            {
                Hashtable notOpened = new Hashtable();
                notOpened[typeof(ConnectionOpenOkBody)] = connectionOpenOk;
                notOpened[typeof(ConnectionCloseBody)] = connectionClose;
                _state2HandlersMap[AMQState.CONNECTION_NOT_OPENED] = notOpened;
            }
            {
                Hashtable open = new Hashtable();
                open[typeof(ChannelCloseBody)] = channelClose;
                open[typeof(ConnectionCloseBody)] = connectionClose;
                open[typeof(BasicDeliverBody)] = basicDeliver;
                open[typeof(BasicReturnBody)] = basicReturn;
                open[typeof(QueueDeleteOkBody)] = queueDeleteOk;
                open[typeof(QueuePurgeOkBody)] = queuePurgeOk;
                _state2HandlersMap[AMQState.CONNECTION_OPEN] = open;
            }
            {
                Hashtable closing = new Hashtable();
                closing[typeof(ConnectionCloseOkBody)] = connectionCloseOk;
                _state2HandlersMap[AMQState.CONNECTION_CLOSING] = closing;
            }
        }

        public AMQState CurrentState
        {
            get
            {
                return _currentState;
            }
        }

        /// <summary>
        /// Changes the state.
        /// </summary>
        /// <param name="newState">The new state.</param>
        /// <exception cref="AMQException">if there is an error changing state</exception>
        public void ChangeState(AMQState newState)
        {
            if (InfoLoggingHack)
            {
                _logger.Debug("State changing to " + newState + " from old state " + _currentState);
            }
            _logger.Debug("State changing to " + newState + " from old state " + _currentState);
            AMQState oldState = _currentState;
            _currentState = newState;

            lock ( _syncLock )
            {
               foreach ( IStateListener l in _stateListeners )
               {
                  l.StateChanged(oldState, newState);
               }
            }
        }

        public void Error(Exception e)
        {
            _logger.Debug("State manager receive error notification: " + e);
            lock ( _syncLock )
            {
               foreach ( IStateListener l in _stateListeners )
               {
                  l.Error(e);
               }
            }
        }

        public bool MethodReceived(AMQMethodEvent evt)
        {
            _logger.Debug(String.Format("Finding method handler. currentState={0} type={1}", _currentState, evt.Method.GetType()));            
            IStateAwareMethodListener handler = FindStateTransitionHandler(_currentState, evt.Method);
            if (handler != null)
            {
                handler.MethodReceived(this, evt);
                return true;
            }
            return false;
        }

        /// <summary>
        /// Finds the state transition handler.
        /// </summary>
        /// <param name="currentState">State of the current.</param>
        /// <param name="frame">The frame.</param>
        /// <returns></returns>
        /// <exception cref="IllegalStateTransitionException">if the state transition if not allowed</exception>
        private IStateAwareMethodListener FindStateTransitionHandler(AMQState currentState,
                                                                     AMQMethodBody frame)                
        {
            Type clazz = frame.GetType();
            if (_logger.IsDebugEnabled)
            {
                _logger.Debug("Looking for state transition handler for frame " + clazz);
            }
            IDictionary classToHandlerMap = (IDictionary) _state2HandlersMap[currentState];

            if (classToHandlerMap == null)
            {
                // if no specialised per state handler is registered look for a
                // handler registered for "all" states
                return FindStateTransitionHandler(AMQState.ALL, frame);
            }
            IStateAwareMethodListener handler = (IStateAwareMethodListener) classToHandlerMap[clazz];
            if (handler == null)
            {
                if (currentState == AMQState.ALL)
                {
                    _logger.Debug("No state transition handler defined for receiving frame " + frame);
                    return null;
                }
                else
                {
                    // if no specialised per state handler is registered look for a
                    // handler registered for "all" states
                    return FindStateTransitionHandler(AMQState.ALL, frame);
                }
            }
            else
            {
                return handler;
            }
        }

        public void AddStateListener(IStateListener listener)
        {
            _logger.Debug("Adding state listener");
            lock ( _syncLock )
            {
               _stateListeners.Add(listener);
            }
        }

        public void RemoveStateListener(IStateListener listener)
        {
           lock ( _syncLock )
           {
              _stateListeners.Remove(listener);
           }
        }

        public void AttainState(AMQState s)
        {
            if (_currentState != s)
            {
                StateWaiter sw = null;
                try
                {
                    _logger.Debug("Adding state wait to reach state " + s);
                    sw = new StateWaiter(s);
                    AddStateListener(sw);
                    sw.WaituntilStateHasChanged();
                    // at this point the state will have changed.
                }
                finally
                { 
                    RemoveStateListener(sw);
                }
            }
        }        
    }
}

File: cotsSource/org.apache.qpid/java/broker/src/main/java/org/apache/qpid/server/state/AMQStateManager.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.server.state;

import java.util.EnumMap;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArraySet;

import org.apache.log4j.Logger;

import org.apache.qpid.AMQException;
import org.apache.qpid.AMQConnectionException;
import org.apache.qpid.framing.*;
import org.apache.qpid.protocol.AMQMethodEvent;
import org.apache.qpid.protocol.AMQMethodListener;
import org.apache.qpid.protocol.AMQConstant;
import org.apache.qpid.server.handler.BasicAckMethodHandler;
import org.apache.qpid.server.handler.BasicCancelMethodHandler;
import org.apache.qpid.server.handler.BasicConsumeMethodHandler;
import org.apache.qpid.server.handler.BasicGetMethodHandler;
import org.apache.qpid.server.handler.BasicPublishMethodHandler;
import org.apache.qpid.server.handler.BasicQosHandler;
import org.apache.qpid.server.handler.BasicRecoverMethodHandler;
import org.apache.qpid.server.handler.BasicRejectMethodHandler;
import org.apache.qpid.server.handler.ChannelCloseHandler;
import org.apache.qpid.server.handler.ChannelCloseOkHandler;
import org.apache.qpid.server.handler.ChannelFlowHandler;
import org.apache.qpid.server.handler.ChannelOpenHandler;
import org.apache.qpid.server.handler.ConnectionCloseMethodHandler;
import org.apache.qpid.server.handler.ConnectionCloseOkMethodHandler;
import org.apache.qpid.server.handler.ConnectionOpenMethodHandler;
import org.apache.qpid.server.handler.ConnectionSecureOkMethodHandler;
import org.apache.qpid.server.handler.ConnectionStartOkMethodHandler;
import org.apache.qpid.server.handler.ConnectionTuneOkMethodHandler;
import org.apache.qpid.server.handler.ExchangeBoundHandler;
import org.apache.qpid.server.handler.ExchangeDeclareHandler;
import org.apache.qpid.server.handler.ExchangeDeleteHandler;
import org.apache.qpid.server.handler.QueueBindHandler;
import org.apache.qpid.server.handler.QueueDeclareHandler;
import org.apache.qpid.server.handler.QueueDeleteHandler;
import org.apache.qpid.server.handler.QueuePurgeHandler;
import org.apache.qpid.server.handler.TxCommitHandler;
import org.apache.qpid.server.handler.TxRollbackHandler;
import org.apache.qpid.server.handler.TxSelectHandler;
import org.apache.qpid.server.protocol.AMQProtocolSession;
import org.apache.qpid.server.virtualhost.VirtualHostRegistry;

/**
 * The state manager is responsible for managing the state of the protocol session. <p/> For each AMQProtocolHandler
 * there is a separate state manager.
 */
public class AMQStateManager implements AMQMethodListener
{
    private static final Logger _logger = Logger.getLogger(AMQStateManager.class);

    private final VirtualHostRegistry _virtualHostRegistry;
    private final AMQProtocolSession _protocolSession;
    /** The current state */
    private AMQState _currentState;

    /**
     * Maps from an AMQState instance to a Map from Class to StateTransitionHandler. The class must be a subclass of
     * AMQFrame.
     */
/*    private final EnumMap<AMQState, Map<Class<? extends AMQMethodBody>, StateAwareMethodListener<? extends AMQMethodBody>>> _state2HandlersMap =
        new EnumMap<AMQState, Map<Class<? extends AMQMethodBody>, StateAwareMethodListener<? extends AMQMethodBody>>>(
            AMQState.class);
  */


    private CopyOnWriteArraySet<StateListener> _stateListeners = new CopyOnWriteArraySet<StateListener>();

    public AMQStateManager(VirtualHostRegistry virtualHostRegistry, AMQProtocolSession protocolSession)
    {

        _virtualHostRegistry = virtualHostRegistry;
        _protocolSession = protocolSession;
        _currentState = AMQState.CONNECTION_NOT_STARTED;

    }

    /*
    protected void registerListeners()
    {
        Map<Class<? extends AMQMethodBody>, StateAwareMethodListener<? extends AMQMethodBody>> frame2handlerMap;

        frame2handlerMap = new HashMap<Class<? extends AMQMethodBody>, StateAwareMethodListener<? extends AMQMethodBody>>();
        _state2HandlersMap.put(AMQState.CONNECTION_NOT_STARTED, frame2handlerMap);

        frame2handlerMap = new HashMap<Class<? extends AMQMethodBody>, StateAwareMethodListener<? extends AMQMethodBody>>();
        _state2HandlersMap.put(AMQState.CONNECTION_NOT_AUTH, frame2handlerMap);

        frame2handlerMap = new HashMap<Class<? extends AMQMethodBody>, StateAwareMethodListener<? extends AMQMethodBody>>();
        _state2HandlersMap.put(AMQState.CONNECTION_NOT_TUNED, frame2handlerMap);

        frame2handlerMap = new HashMap<Class<? extends AMQMethodBody>, StateAwareMethodListener<? extends AMQMethodBody>>();
        frame2handlerMap.put(ConnectionOpenBody.class, ConnectionOpenMethodHandler.getInstance());
        _state2HandlersMap.put(AMQState.CONNECTION_NOT_OPENED, frame2handlerMap);

        //
        // ConnectionOpen handlers
        //
        frame2handlerMap = new HashMap<Class<? extends AMQMethodBody>, StateAwareMethodListener<? extends AMQMethodBody>>();
        ChannelOpenHandler.getInstance();
        ChannelCloseHandler.getInstance();
        ChannelCloseOkHandler.getInstance();
        ConnectionCloseMethodHandler.getInstance();
        ConnectionCloseOkMethodHandler.getInstance();
        ConnectionTuneOkMethodHandler.getInstance();
        ConnectionSecureOkMethodHandler.getInstance();
        ConnectionStartOkMethodHandler.getInstance();
        ExchangeDeclareHandler.getInstance();
        ExchangeDeleteHandler.getInstance();
        ExchangeBoundHandler.getInstance();
        BasicAckMethodHandler.getInstance();
        BasicRecoverMethodHandler.getInstance();
        BasicConsumeMethodHandler.getInstance();
        BasicGetMethodHandler.getInstance();
        BasicCancelMethodHandler.getInstance();
        BasicPublishMethodHandler.getInstance();
        BasicQosHandler.getInstance();
        QueueBindHandler.getInstance();
        QueueDeclareHandler.getInstance();
        QueueDeleteHandler.getInstance();
        QueuePurgeHandler.getInstance();
        ChannelFlowHandler.getInstance();
        TxSelectHandler.getInstance();
        TxCommitHandler.getInstance();
        TxRollbackHandler.getInstance();
        BasicRejectMethodHandler.getInstance();

        _state2HandlersMap.put(AMQState.CONNECTION_OPEN, frame2handlerMap);

        frame2handlerMap = new HashMap<Class<? extends AMQMethodBody>, StateAwareMethodListener<? extends AMQMethodBody>>();

        _state2HandlersMap.put(AMQState.CONNECTION_CLOSING, frame2handlerMap);

    } */

    public AMQState getCurrentState()
    {
        return _currentState;
    }

    public void changeState(AMQState newState) throws AMQException
    {
        _logger.debug("State changing to " + newState + " from old state " + _currentState);
        final AMQState oldState = _currentState;
        _currentState = newState;

        for (StateListener l : _stateListeners)
        {
            l.stateChanged(oldState, newState);
        }
    }

    public void error(Exception e)
    {
        _logger.error("State manager received error notification[Current State:" + _currentState + "]: " + e, e);
        for (StateListener l : _stateListeners)
        {
            l.error(e);
        }
    }

    public <B extends AMQMethodBody> boolean methodReceived(AMQMethodEvent<B> evt) throws AMQException
    {
        MethodDispatcher dispatcher = _protocolSession.getMethodDispatcher();

        final int channelId = evt.getChannelId();
        B body = evt.getMethod();

        if(channelId != 0 && _protocolSession.getChannel(channelId)== null)
        {

            if(! ((body instanceof ChannelOpenBody)
                  || (body instanceof ChannelCloseOkBody)
                  || (body instanceof ChannelCloseBody)))
            {
                throw body.getConnectionException(AMQConstant.CHANNEL_ERROR, "channel is closed");
            }

        }

        return body.execute(dispatcher, channelId);

    }

    private <B extends AMQMethodBody> void checkChannel(AMQMethodEvent<B> evt, AMQProtocolSession protocolSession)
        throws AMQException
    {
        if ((evt.getChannelId() != 0) && !(evt.getMethod() instanceof ChannelOpenBody)
                && (protocolSession.getChannel(evt.getChannelId()) == null)
                && !protocolSession.channelAwaitingClosure(evt.getChannelId()))
        {
            throw evt.getMethod().getChannelNotFoundException(evt.getChannelId());
        }
    }

/*
    protected <B extends AMQMethodBody> StateAwareMethodListener<B> findStateTransitionHandler(AMQState currentState,
        B frame)
    // throws IllegalStateTransitionException
    {
        final Map<Class<? extends AMQMethodBody>, StateAwareMethodListener<? extends AMQMethodBody>> classToHandlerMap =
            _state2HandlersMap.get(currentState);

        final StateAwareMethodListener<B> handler =
            (classToHandlerMap == null) ? null : (StateAwareMethodListener<B>) classToHandlerMap.get(frame.getClass());

        if (handler == null)
        {
            _logger.debug("No state transition handler defined for receiving frame " + frame);

            return null;
        }
        else
        {
            return handler;
        }
    }
*/

    public void addStateListener(StateListener listener)
    {
        _logger.debug("Adding state listener");
        _stateListeners.add(listener);
    }

    public void removeStateListener(StateListener listener)
    {
        _stateListeners.remove(listener);
    }

    public VirtualHostRegistry getVirtualHostRegistry()
    {
        return _virtualHostRegistry;
    }

    public AMQProtocolSession getProtocolSession()
    {
        return _protocolSession;
    }
}
